paint/data/dataset.py

```diff
@@ -0,0 +1,437 @@
+                root_dir=root_dir, item_type=item_type, heliostats=heliostats
+            )
+        else:
+            # Check if the folder exists, if not the data must be downloaded.
+            if not root_dir.is_dir():
+                log.warning(
+                    "The root directory does not exist, the data has not yet been downloaded!\n"
+                    "The data will now be downloaded..."
+                )
+                cls._download_heliostat_data(
+                    root_dir=root_dir, item_type=item_type, heliostats=heliostats
+                )
+        return cls(
+            root_dir=Path(root_dir),
+            item_type=item_type,
+        )
+
+    @staticmethod
+    def _download_heliostat_data(
+        root_dir: Union[str, Path],
+        item_type: str,
+        heliostats: Union[list[str], None] = None,
+    ):
+        """
+        Download the heliostat calibration data for the dataset based on heliostats.
+
+        Parameters
+        ----------
+        root_dir : Union[str, Path]
+            Directory where the data will be stored.
+        item_type : str
+            Type of item being downloaded, i.e. raw image, cropped image, flux image, or calibration properties.
+        heliostats : Union[list[str], None]
+            List of heliostats for which calibration data should be downloaded. If no list is provided the data for all
+            heliostats will be downloaded (Default: ``None``).
+        """
+        # Create STAC client.
+        client = StacClient(output_dir=root_dir)
+        client.get_heliostat_data(
+            heliostats=heliostats,
+            collections=[mappings.SAVE_CALIBRATION.lower()],
+            filtered_calibration_keys=[item_type],
+            for_dataset=True,
+        )
+
+    def __len__(self) -> int:
+        """
+        Calculate the length of the dataset.
+
+        Returns
+        -------
+        int
+            Length of the dataset.
+        """
+        return len(self.item_ids)
+
+    def __getitem__(self, idx: int) -> Union[torch.Tensor, dict[str, Any]]:
+        """
+        Return the item at position ``idx``.
+
+        Parameters
+        ----------
+        idx : int
+            Index of the item to be retrieved.
+
+        Returns
+        -------
+        Union[torch.Tensor, dict[str, Any]]
+            Either the image as a torch.Tensor or a dict containing the calibration properties, depending on the item
+            type selected.
+        """
+        id_ = self.item_ids[idx]
+        file_name = self.root_dir / f"{id_}{self.file_identifier}"
+        if self.file_identifier in [
+            mappings.CALIBRATION_RAW_IMAGE_IDENTIFIER,
+            mappings.CALIBRATION_CROPPED_IMAGE_IDENTIFIER,
+            mappings.CALIBRATION_FLUX_IMAGE_IDENTIFIER,
+            mappings.CALIBRATION_FLUX_CENTERED_IMAGE_IDENTIFIER,
+        ]:
+            item = cv2.imread(str(file_name))
+            item = self.to_tensor(item)
```

Comment:
Do you want to use the torchvision.ToTensor() transform here? This might avoid double allocation of memory

Reply:
Done